File: Leetcode/1020 Number of Enclaves.ts
function numEnclaves(grid: number[][]): number {
    const n = grid.length, m = grid[0].length;
    let enclaves = 0
    for(let i = 1; i < n - 1; i++) {
        for(let j = 1; j < m - 1;j++) {
            if(grid[i][j] == 1) {
                enclaves += Math.max(0, dfs(i, j, grid))
            }
        }
    }
    return enclaves;
}

function dfs(i: number,j: number, grid: number[][]) {
    const n = grid.length, m = grid[0].length;
    // touched edge so this whole land mass is not valid
    if(i < 0 || i == n || j < 0 || j == m) {
        return -1; 
    }
    if(grid[i][j] != 1) {
        return 0;
    }
    grid[i][j] = -1;
    const neighbourCounts = [
        dfs(i - 1, j, grid), 
        dfs(i, j + 1, grid), 
        dfs(i + 1, j, grid), 
        dfs(i, j - 1, grid)
    ];
    let totalCount = 1;
    for(const neighbourCount of neighbourCounts) {
        // any neighbout is invalid the whole lang mass is invalid
        if(neighbourCount < 0) {
            return -1;
        }
        totalCount += neighbourCount;
    }
    return totalCount;
}

File: Leetcode/1020 Number of Enclaves.py
class Solution:
    def numEnclaves(self, grid):
        n, m = len(grid), len(grid[0])
        for i in range(n):
            if grid[i][0] == 1:
                dfs(i, 0, grid)
            if grid[i][m-1] == 1:
                dfs(i, m-1, grid)
        for j in range(m):
            if grid[0][j] == 1:
                dfs(0, j, grid)
            if grid[n-1][j] == 1:
                dfs(n-1, j, grid)
        enclaves = 0
        for row in grid:
            for cell in row:
                if cell == 1:
                    enclaves += 1
        return enclaves


def dfs(i, j, grid):
    n, m = len(grid), len(grid[0])
    if i < 0 or i == n or j < 0 or j == m or grid[i][j] != 1:
        return
    grid[i][j] = -1
    dfs(i - 1, j, grid)
    dfs(i, j + 1, grid)
    dfs(i + 1, j, grid)
    dfs(i, j - 1, grid)
